### DISCUSSION TITLE
Build message about missing Extensibility Essentials extension?

### DISCUSSION THREAD

**Discussion by reduckted** [VsixCommunity/Community.VisualStudio.Toolkit#13] [Ideas]
This could just be a crazy idea, but what does everyone think about this library producing a build message if the _Extensibility Essentials_ extension isn't installed? Something like this:

![image](https://user-images.githubusercontent.com/10321525/115388541-580f0f80-a21f-11eb-9af4-b5e3aeb6a718.png)

It would help newcomers who may not be aware of that extension, and you could also use it for your own open source extensions to alert contributors to that extension.

I've put a prototype together, and it's definitely possible, but I'm just not sure if it's something that this library should do.

Here's how it works.
1. The _Extensibility Essentials_ extension uses the `IVsBuildPropertyStorage` service to add a build property to every project when a solution is opened.
2. A `.targets` file in `Community.VisualStudio.Toolkit` would define a build target that looks for the presence of that build property (and a build property that indicates it's running inside Visual Studio so that you don't produce the message for command line builds).
3. If the build property is not found, then the extension is not installed and the target would produce a message (or warning/error if you wanted to enforce it).

So, good idea? Bad idea? What does everyone think?

**Comment by madskristensen** [comment]
Fantastic idea!!

A couple of things:

1. Let's only add it to VSIX projects instead of every project (just to play it safe)
2. How do you turn off the warning if you don't want it?

**Reply by reduckted** [reply]
> Let's only add it to VSIX projects instead of every project (just to play it safe)

That would be projects with a `.vsixmanifest` file, right? Or is there a better way to detect that? I guess we could possibly look for the `Community.VisualStudio.Toolkit` package reference, though I'm not sure how difficult that would be.

> How do you turn off the warning if you don't want it?

In my prototype you could configure it to produce either nothing (disabled) an info message, a warning or an error by setting a property in the project file:
```xml
<PropertyGroup>
    <ExtensibilityExtensionsMessageLevel>error</ExtensibilityExtensionsMessageLevel> <!-- or "warning", "info" or "none" -->
</PropertyGroup>
```

**Reply by madskristensen** [reply]
VSIX projects have their own subtype GUID specified in the .csproj, which is `82b43b9b-a64c-4715-b499-d71e9ca2bd60`

**Comment by yannduran** [comment]
I'm so happy that someone else has thought to bring this up. I had this on a list of things to bring up, but the list keeps getting longer, and time seems to just get shorter.

I was also concerned that if people didn't at least have the VSIX/VSCT extensions installed, that the Toolkit would have been of far less use, or even worse, confusing to newcomers.

I agree with Mads that it should only be added to the VSIX project. I'm not sure I'd worry about how to turn it off unless it seems surfaces as a problem that actually needs solving.
